src/fsharp/CheckDeclarations.fs

```diff
@@ -0,0 +1,5888 @@
+        { env with eNameResEnv = AddTyconRefsToNameEnv BulkAdd.Yes false g amap env.eAccessRights scopem true env.eNameResEnv tcrefs }
+    env
+
+/// Adjust the TcEnv to account for a fully processed "namespace" declaration in this file
+let AddLocalRootModuleOrNamespace tcSink g amap scopem env (mtyp: ModuleOrNamespaceType) = 
+    // Compute the top-rooted module or namespace references
+    let modrefs = mtyp.ModuleAndNamespaceDefinitions |> List.map mkLocalModRef
+    // Compute the top-rooted type definitions
+    let tcrefs = mtyp.TypeAndExceptionDefinitions |> List.map mkLocalTyconRef
+    let env = AddRootModuleOrNamespaceRefs g amap scopem env modrefs
+    let env = { env with
+                    eNameResEnv = if isNil tcrefs then env.eNameResEnv else AddTyconRefsToNameEnv BulkAdd.No false g amap env.eAccessRights scopem true env.eNameResEnv tcrefs
+                    eUngeneralizableItems = addFreeItemOfModuleTy mtyp env.eUngeneralizableItems }
+    CallEnvSink tcSink (scopem, env.NameEnv, env.eAccessRights)
+    env
+
+/// Inside "namespace X.Y.Z" there is an implicit open of "X.Y.Z"
+let ImplicitlyOpenOwnNamespace tcSink g amap scopem enclosingNamespacePath (env: TcEnv) = 
+    if isNil enclosingNamespacePath then 
+        env
+    else
+        // For F# interactive, skip "FSI_0002" prefixes when determining the path to open implicitly
+        let enclosingNamespacePathToOpen = 
+            match TryStripPrefixPath g enclosingNamespacePath with 
+            | Some(_, rest) -> rest
+            | None -> enclosingNamespacePath
+
+        match enclosingNamespacePathToOpen with
+        | id :: rest ->
+            let ad = env.AccessRights
+            match ResolveLongIdentAsModuleOrNamespace tcSink ResultCollectionSettings.AllResults amap scopem true OpenQualified env.eNameResEnv ad id rest true with 
+            | Result modrefs -> 
+                let modrefs = List.map p23 modrefs
+                let openTarget = SynOpenDeclTarget.ModuleOrNamespace(enclosingNamespacePathToOpen, scopem)
+                let openDecl = OpenDeclaration.Create (openTarget, modrefs, [], scopem, true)
+                OpenModuleOrNamespaceRefs tcSink g amap scopem false env modrefs openDecl
+            | Exception _ -> env
+        | _ -> env
+
+//-------------------------------------------------------------------------
+// Bind elements of data definitions for exceptions and types (fields, etc.)
+//------------------------------------------------------------------------- 
+
+exception NotUpperCaseConstructor of range
+
+let CheckNamespaceModuleOrTypeName (g: TcGlobals) (id: Ident) = 
+    // type names '[]' etc. are used in fslib
+    if not g.compilingFslib && id.idText.IndexOfAny IllegalCharactersInTypeAndNamespaceNames <> -1 then 
+        errorR(Error(FSComp.SR.tcInvalidNamespaceModuleTypeUnionName(), id.idRange))
+
+let CheckDuplicates (idf: _ -> Ident) k elems = 
+    elems |> List.iteri (fun i uc1 -> 
```

Comment:
I suggest a small optimization here:
```
let CheckDuplicates (idf: _ -> Ident) k elems = 
    let ids = elems |> List.mapi (fun i uc -> i, idf uc)
    ids
        |> List.iter (fun (i, id1) ->
            ids
                |> List.iter (fun (j, id2) ->
                    if j > i && id1.idText = id2.idText then 
                        errorR (Duplicate(k, id1.idText, id1.idRange))))
    elems
```
because is computing idf for all elements too many times ```n * (n + 1)``` instead of just ```n```

Reply:
could you please send it as new PR. I'd also suggest you unroll it as nested for loops .

Reply:
The suggested code will look like this:
```
let CheckDuplicates (idf: _ -> Ident) k elems = 
    let ids = elems |> List.mapi (fun i uc -> i, idf uc)
    for (i, id1) in ids do
        for (j, id2) in ids do
            if j > i && id1.idText = id2.idText then 
                errorR (Duplicate(k, id1.idText, id1.idRange))
    elems
```

Reply:
@forki The PR is https://github.com/dotnet/fsharp/pull/10325